batch2/batch/driver/instance.py

```diff
@@ -1,273 +1,102 @@
-            log.info(f'deactivating instance {self.name}')
-            start = time.time()
-            if self.pending:
-                self.pending = False
-                self.inst_pool.n_pending_instances -= 1
-                self.inst_pool.free_cores_mcpu -= self.inst_pool.worker_capacity_mcpu
-                assert not self.active
-                log.info(f'{self.inst_pool.n_pending_instances} pending {self.inst_pool.n_active_instances} active workers')
-                return
-
-            if not self.active:
-                return
-
-            self.mark_as_unhealthy()
-
-            pod_list = list(self.pods)
-            await asyncio.gather(*[p.unschedule() for p in pod_list])
-            assert not self.pods
-
-            for pod in pod_list:
-                asyncio.ensure_future(pod.put_on_ready())
-
-            self.active = False
-
-            log.info(f'took {time.time() - start} seconds to deactivate {self.name}')
-            log.info(f'{self.inst_pool.n_pending_instances} pending {self.inst_pool.n_active_instances} active workers')
-
-    def update_timestamp(self):
-        if self in self.inst_pool.instances:
-            self.inst_pool.instances.remove(self)
-            self.last_updated = time.time()
-            self.inst_pool.instances.add(self)
-
-    def mark_as_unhealthy(self):
-        if not self.active or not self.healthy:
-            return
-
-        self.inst_pool.instances.remove(self)
-        self.healthy = False
-        self.inst_pool.instances.add(self)
-
-        if self in self.inst_pool.instances_by_free_cores:
-            self.inst_pool.instances_by_free_cores.remove(self)
-            self.inst_pool.n_active_instances -= 1
-            self.inst_pool.free_cores_mcpu -= self.free_cores_mcpu
-
-        self.update_timestamp()
-
-    def mark_as_healthy(self):
-        self.last_ping = time.time()
-
-        if not self.active or self.healthy:
+        if self._state in ('inactive', 'deleted'):
             return
 
-        self.inst_pool.instances.remove(self)
-        self.healthy = True
-        self.inst_pool.instances.add(self)
-
-        if self not in self.inst_pool.instances_by_free_cores:
-            self.inst_pool.n_active_instances += 1
-            self.inst_pool.instances_by_free_cores.add(self)
-            self.inst_pool.free_cores_mcpu += self.free_cores_mcpu
-            self.inst_pool.driver.changed.set()
+        await check_call_procedure(
+            self.db,
+            'CALL deactivate_instance(%s);',
+            (self.id,))
 
-    async def remove(self):
-        log.info(f'removing instance {self.name}')
-        await self.deactivate()
-        self.inst_pool.instances.remove(self)
-        if self.token in self.inst_pool.token_inst:
-            del self.inst_pool.token_inst[self.token]
-        await self.inst_pool.driver.db.instances.delete_record(self.name)
+        self.instance_pool.adjust_for_remove_instance(self)
+        self._state = 'inactive'
```

Comment:
Correct!  So cancel overall looks like:
 - Mark the batch cancelled in the front end, and call the driver /cancel endpoint which sets cancelled_state_chagned
 - That wakes up the cancel loop in the scheduler, which finds cancelled, running jobs, and call mark_unscheduled.  Jobs become ready.
 - Ready, cancelled jobs (whether unscheduled or not-yet-scheduled) will get scheduled as normal, and the first test in the scheduling loop will mark them as cancelled instead of running them.